### PR TITLE
Improve implementations of addAllObjectsToIndex

### PR DESCRIPTION
**Pull request by Beacze**
- move implementation of addAllObjectToIndex to SearchService
- use generics for Type
- move implementations of removeFromIndex, saveToIndex and findAll to SearchService

Preparation for usage of asynchronous indexing.

### CODE REVIEW

File: Kitodo/src/main/java/org/kitodo/services/data/base/SearchService.java [hunk 1]
```diff
@@ -57,29 +58,34 @@
...
+     * frontend.
      *
-     * @param baseIndexedBean
-     *            object
+     * @param jsonObject
+     *            return from find methods
+     * @return DTO object
      */
-    public abstract void saveToIndex(T baseIndexedBean) throws CustomResponseException, IOException;
+    public abstract S convertJSONObjectToDTO(JSONObject jsonObject, boolean related) throws DataException;
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/services/data/base/SearchService.java]: Missing JavaDoc for parameter `related`.

File: Kitodo/src/main/java/org/kitodo/services/data/base/SearchService.java [hunk 2]
```diff
@@ -57,29 +58,34 @@
...
+    protected BaseType type;
 
     /**
      * Constructor necessary to use searcher in child classes.
      *
      * @param searcher
      *            for executing queries
      */
-    public SearchService(V dao, Searcher searcher) {
+    public SearchService(V dao, BaseType type, Indexer indexer, Searcher searcher) {
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/services/data/base/SearchService.java]: Missing JavaDoc for parameter `dao`, `type` and `indexer`.

File: Kitodo/src/main/java/org/kitodo/services/data/base/SearchService.java (hunk 1 above)
Comment (Beacze) [reply Kitodo/src/main/java/org/kitodo/services/data/base/SearchService.java]: added

File: Kitodo/src/main/java/org/kitodo/services/data/base/SearchService.java (hunk 2 above)
Comment (Beacze) [reply Kitodo/src/main/java/org/kitodo/services/data/base/SearchService.java]: added

File: Kitodo/src/main/java/org/kitodo/services/data/base/TitleSearchService.java [hunk 3]
```diff
@@ -28,10 +30,11 @@
  * Class for implementing methods used by service classes which search for title
  * in ElasticSearch index.
  */
-public abstract class TitleSearchService<T extends BaseIndexedBean, S extends BaseDTO, V extends BaseDAO<T>> extends SearchService<T, S, V> {
+public abstract class TitleSearchService<T extends BaseIndexedBean, S extends BaseDTO, V extends BaseDAO<T>>
+        extends SearchService<T, S, V> {
 
-    public TitleSearchService(V dao, Searcher searcher) {
-        super(dao, searcher);
+    public TitleSearchService(V dao, BaseType type, Indexer indexer, Searcher searcher) {
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/services/data/base/TitleSearchService.java]: Missing JavaDoc for class constructor or add a override annotation.

Comment (Beacze) [reply Kitodo/src/main/java/org/kitodo/services/data/base/TitleSearchService.java]: done

**Pull request merged**